File: src/rust/src/hclust.rs
use kodama::{linkage, Method, Step};

/// Resolve a linkage method by name.
///
/// Names are validated on the R side; an unknown name here is a bug.
///
/// Note that kodama squares the dissimilarities for Ward, centroid and median
/// and takes the square root afterwards, matching fastcluster and SciPy. Ward
/// therefore corresponds to R's `"ward.D2"`, and centroid and median take plain
/// distances rather than the squared distances `stats::hclust()` expects.
pub fn method_from_name(name: &str) -> Method {
    match name {
        "single" => Method::Single,
        "complete" => Method::Complete,
        "average" => Method::Average,
        "weighted" => Method::Weighted,
        "ward" => Method::Ward,
        "centroid" => Method::Centroid,
        "median" => Method::Median,
        _ => panic!("Unknown linkage method: {name}"),
    }
}

/// Translate a kodama cluster label into R's `hclust` merge encoding.
///
/// kodama labels observations `0..n` and the cluster formed at step `i` as
/// `n + i` (SciPy's scheme). R encodes a singleton observation as its negated
/// 1-based index, and a previously formed cluster as the positive 1-based
/// index of the step that created it.
#[inline]
fn to_merge_entry(label: usize, n: usize) -> i32 {
    if label < n {
        -((label + 1) as i32)
    } else {
        (label - n + 1) as i32
    }
}

/// Compute the leaf ordering R uses to draw a dendrogram without crossings.
///
/// Depth-first from the root merge, emitting observations as they are reached.
/// Iterative rather than recursive so deep trees cannot overflow the stack.
fn leaf_order(steps: &[Step<f64>], n: usize) -> Vec<i32> {
    let mut order = Vec::with_capacity(n);

    if steps.is_empty() {
        if n == 1 {
            order.push(1);
        }
        return order;
    }

    // The root is the final step, in R's 1-based step numbering.
    let mut stack: Vec<i32> = vec![steps.len() as i32];

    while let Some(node) = stack.pop() {
        if node < 0 {
            order.push(-node);
        } else {
            let step = &steps[(node - 1) as usize];
            // Push the right child first so the left is popped, and emitted, first.
            stack.push(to_merge_entry(step.cluster2, n));
            stack.push(to_merge_entry(step.cluster1, n));
        }
    }

    order
}

/// The pieces of an `hclust` object.
pub struct Hclust {
    /// `(n - 1) x 2`, column-major, in R's merge encoding.
    pub merge: Vec<i32>,
    pub height: Vec<f64>,
    pub order: Vec<i32>,
}

/// Run hierarchical clustering over a condensed dissimilarity matrix.
///
/// `condensed` is consumed: kodama's `linkage` takes it by mutable reference and
/// destroys it while working, so callers must pass a copy they do not need.
pub fn hclust(mut condensed: Vec<f64>, n: usize, method: Method) -> Hclust {
    let dend = linkage(&mut condensed, n, method);
    let steps = dend.steps();

    let n_steps = steps.len();
    let mut merge = vec![0i32; 2 * n_steps];
    let mut height = Vec::with_capacity(n_steps);

    for (k, step) in steps.iter().enumerate() {
        // R matrices are column-major: (k, 0) at k, (k, 1) at n_steps + k.
        merge[k] = to_merge_entry(step.cluster1, n);
        merge[n_steps + k] = to_merge_entry(step.cluster2, n);
        height.push(step.dissimilarity);
    }

    Hclust {
        merge,
        height,
        order: leaf_order(steps, n),
    }
}

File: R/hclust.R
#' Hierarchical Agglomerative Clustering
#'
#' Agglomerative clustering via the Rust
#' \href{https://github.com/diffeo/kodama}{kodama} crate, a port of
#' \emph{fastcluster}. Returns a standard [stats::hclust] object, so
#' [stats::cutree()], [stats::as.dendrogram()], `plot()` and the rest of R's
#' hierarchical clustering ecosystem work on the result unchanged.
#'
#' # Linkage methods
#'
#' Dissimilarities are squared internally for `"ward"`, `"centroid"` and
#' `"median"`, with the square root taken afterwards, the fastcluster and SciPy
#' convention. Two consequences differ from [stats::hclust()]:
#'
#' - `"ward"` corresponds to R's `"ward.D2"`, not `"ward.D"`.
#' - `"centroid"` and `"median"` take plain distances here, whereas
#'   [stats::hclust()] expects them to be squared beforehand.
#'
#' `"weighted"` is R's `"mcquitty"` (WPGMA); `"average"` is UPGMA.
#'
#' `"centroid"` and `"median"` can produce inversions, a merge at a lower
#' height than the one before it. This is a property of the methods, not a bug,
#' but [stats::cutree()] rejects such trees, so a warning is issued when it
#' happens.
#'
#' @param d A [stats::dist] object, or a numeric matrix or data frame, in which
#'   case Euclidean distances are computed with [shoal_dist()] first.
#' @param method Linkage method. One of `"complete"`, `"single"`, `"average"`,
#'   `"weighted"`, `"ward"`, `"centroid"` or `"median"`.
#'
#' @returns An object of class `"hclust"` with components `merge`, `height`,
#'   `order`, `labels`, `method`, `call` and `dist.method`.
#'
#' @seealso [shoal_dist()] for building the distance matrix.
#'
#' @examples
#' d <- shoal_dist(as.matrix(iris[, 1:4]))
#' fit <- shoal_hclust(d, method = "ward")
#' cutree(fit, k = 3)
#'
#' @export
shoal_hclust <- function(d,
                         method = c(
                           "complete", "single", "average", "weighted",
                           "ward", "centroid", "median"
                         )) {
  method <- rlang::arg_match(method)

  if (!inherits(d, "dist")) {
    if (looks_like_dist_matrix(d)) {
      cli::cli_abort(c(
        "{.arg d} looks like a square distance matrix, not raw data.",
        "i" = "Pass {.code as.dist(d)} to use it as distances."
      ))
    }
    d <- shoal_dist(d)
  }

  n <- attr(d, "Size")
  if (is.null(n) || n < 2L) {
    cli::cli_abort("{.arg d} must describe at least 2 observations.")
  }

  # In double: the integer product overflows past 46,341 observations.
  expected <- as.double(n) * (n - 1) / 2
  if (length(d) != expected) {
    cli::cli_abort(
      "{.arg d} has {length(d)} dissimilarit{?y/ies}, expected {expected} for {n} observations."
    )
  }
  # NA is screened here, cheaply; Inf is caught on the Rust side in the same
  # pass that copies `d`, with the same message. A separate is.finite() pass
  # would allocate a logical vector as long as `d`.
  if (anyNA(d)) {
    cli::cli_abort("{.arg d} must not contain missing or non-finite values.")
  }
  if (!is.double(d)) {
    d <- as.double(d)
  }

  # The dist object goes across as is; the copy kodama needs is made once
  # on the Rust side rather than here as well.
  res <- rust_hclust(d, as.integer(n), method)

  if (is.unsorted(res$height)) {
    cli::cli_warn(c(
      "The dendrogram contains inversions.",
      "i" = "{.val {method}} linkage does not guarantee monotone merge heights.",
      "i" = "{.fn cutree} will reject this tree."
    ))
  }

  structure(
    list(
      merge = res$merge,
      height = res$height,
      order = res$order,
      labels = attr(d, "Labels"),
      method = method,
      call = match.call(),
      dist.method = attr(d, "method")
    ),
    class = "hclust"
  )
}
